### PR TITLE
Use GDPR compliant marketing consent model in Edit Profile

### PR DESCRIPTION
**Pull request by mariogalic**
## What does this change?

_NOTE: Do not merge! This PR might be open for a long time as requirements are evolving, however I would like to collect some initial feedback._

Use new GDPR compliant marketing consent model on the Edit Profile Privacy Tab:
https://github.com/guardian/identity/pull/696

This PR also attempts to rediscover how Edit Profile was originally architected so there is quite a bit of added comments and renaming of methods to hopefully make intent clearer.

Main changes are in:
* `EditProfileController.scala`
* `PrivacyMapping.scala`
* `privacyForm.scala.html`
* `UserFormMapping.scala`

## What is the value of this and can you measure success?

GDPR compliance. 

## Does this affect other platforms - Amp, Apps, etc?

No

## Screenshots

TODO

## Tested in CODE?

Yes

### CODE REVIEW

File: identity/app/controllers/EditProfileController.scala [hunk 1]
```diff
@@ -47,53 +51,84 @@ class EditProfileController(idUrlBuilder: IdentityUrlBuilder,
...
-      val forms = ProfileForms(user, activePage).bindFromRequest(request)
-
-      forms.activeForm.value.map {
-        case data: AccountFormData if (data.deleteTelephone) => {
-          identityApiClient.deleteTelephone(user.auth) map {
-            case Left(errors) => profileFormsView(page, forms.withErrors(errors), user)
+  private def submitForm(page: IdentityPage): Action[AnyContent] =
+    csrfCheck {
+      authActionWithUser.async { implicit request =>
+        def identityPageToEditProfilePage(identityPage: IdentityPage): EditProfilePage =
```
Comment (jfsoul) [inline identity/app/controllers/EditProfileController.scala]: This method seems a little out of place.  The question I have is, do we even need it?  Why can't `EditProfilePage` inherit from `IdentityPage` and its children define the values of `IdentityPage` fields.  E.g.  
```
sealed trait EditProfilePage extends IdentityPage
case object AccountEditProfilePage extends EditProfilePage("/public/edit", "Edit Public Profile")
```

and then there's no need for a mapping?

File: identity/app/form/PrivacyMapping.scala [hunk 2]
```diff
@@ -1,51 +1,81 @@
...
+    *
+    * @param userFromApi Identity User domain model from IDAPI defiend in identity-model library
+    * @return form processing DTO PrivacyFromData
+    */
+  def apply(userFromApi: User): PrivacyFormData =
+    PrivacyFormData(
+      receiveGnmMarketing = userFromApi.statusFields.receiveGnmMarketing.getOrElse(false),
+      receive3rdPartyMarketing = userFromApi.statusFields.receive3rdPartyMarketing.getOrElse(false),
+      allowThirdPartyProfiling = userFromApi.statusFields.allowThirdPartyProfiling.getOrElse(true),
+      consents = if (userFromApi.consents.isEmpty) defaultConsents else userFromApi.consents
```
Comment (jfsoul) [inline identity/app/form/PrivacyMapping.scala]: I kind of thought that `Consents` would be an `Option`, but I suppose that depends on how it is handled by the model.

File: identity/app/idapiclient/UserUpdateDTO.scala [hunk 3]
```diff
@@ -11,5 +17,6 @@ case class UserUpdate (
   password: Option[String] = None,
   userGroups: Option[Set[GroupMembership]] = None,
   socialLinks: Option[Set[SocialLink]] = None,
-  adData: Option[Map[String, AnyRef]] = None
-);
+  adData: Option[Map[String, AnyRef]] = None,
+  consents: Option[List[Consent]] = None
```
Comment (jfsoul) [inline identity/app/idapiclient/UserUpdateDTO.scala]: Are `statusFields` still needed in this now that we have `consents`?

File: identity/app/form/PrivacyMapping.scala [hunk 4]
```diff
@@ -1,51 +1,81 @@
...
+import play.api.i18n.MessagesProvider
 
 class PrivacyMapping extends UserFormMapping[PrivacyFormData] {
 
-  protected def formMapping(implicit messagesProvider: MessagesProvider): Mapping[PrivacyFormData] = mapping(
-    "receiveGnmMarketing" -> boolean,
-    "receive3rdPartyMarketing" -> boolean,
-    "allowThirdPartyProfiling" -> boolean
-  )(PrivacyFormData.apply)(PrivacyFormData.unapply)
+  private val dateTimeFormatISO8601: String = "yyyy-MM-dd'T'HH:mm:ssZZ"
```
Comment (jfsoul) [inline identity/app/form/PrivacyMapping.scala]: get rid of this and replace the usage below with `ISODateTimeFormat.dateTimeNoMillis`? I think they are the same thing.

File: project/ProjectSettings.scala [hunk 5]
```diff
@@ -62,6 +62,7 @@ object ProjectSettings {
     resolvers ++= Seq(
       Resolver.typesafeRepo("releases"),
       Resolver.sonatypeRepo("releases"),
+      Resolver.sonatypeRepo("snapshots"), // FIXME: Is it OK to keep this?
```
Comment (jfsoul) [inline project/ProjectSettings.scala]: I think no is the answer, but you already know that 😄

File: identity/app/views/fragments/profile/privacyForm.scala.html [hunk 6]
```diff
@@ -1,34 +1,67 @@
...
+            '_label -> getConsentText(consentField)
+        )
+    )(nonInputFields, messages)
+}
+
+@v2consentCheckboxes = {
+  @helper.repeat(privacyForm("consents"), min=1) { consentField => @v2ConsentCheckbox(consentField) }
+}
+
+@getConsentText(consentField: Field) = {
```
Comment (jfsoul) [inline identity/app/views/fragments/profile/privacyForm.scala.html]: can we take this helper out of the templates?

File: identity/app/views/fragments/profile/privacyForm.scala.html [hunk 7]
```diff
@@ -1,34 +1,67 @@
...
+        <div class="form__error">
+          Error processing the form. Your changes have not been saved:
+          <p>@privacyForm.errors.map(formError => (formError.key, formError.message)).mkString(", ")</p>
+        </div>
     }
 
+    @* Newsletters *@
     <fieldset class="fieldset">
         <div class="fieldset__heading">
-            <h2 class="form__heading">Guardian & Observer newsletters</h2>
```
Comment (jfsoul) [inline identity/app/views/fragments/profile/privacyForm.scala.html]: Are these changes related?

File: identity/app/controllers/EditProfileController.scala [hunk 8]
```diff
@@ -47,53 +51,84 @@ class EditProfileController(idUrlBuilder: IdentityUrlBuilder,
...
+            case _ => PrivacyEditProfilePage
+          }
 
-            case Right(_) => {
-              val boundForms = forms.bindForms(user.user.copy(privateFields = user.user.privateFields.copy(telephoneNumber = None)))
-              profileFormsView(page, boundForms, user)
+        val activePage = identityPageToEditProfilePage(page)
+        val userDO = request.user
+        val boundProfileForms =
+          ProfileForms(userDO, activePage).bindFromRequestWithAddressErrorHack(request) // NOTE: only active form is bound to request data
```
[the remark this answers is not in the document]
Comment (mario-galic) [reply identity/app/controllers/EditProfileController.scala]: I've inlined it.

File: identity/app/form/PrivacyMapping.scala (hunk 4 above)
Comment (mario-galic) [reply identity/app/form/PrivacyMapping.scala]: Indeed they are the same, however `ISODateTimeFormat.dateTimeNoMillis` returns `DateTimeFormatter` and it seems there is no way to get the underlying pattern string:
https://stackoverflow.com/questions/28949179/get-original-pattern-string-given-a-jdk-8-datetimeformatter

File: identity/app/idapiclient/UserUpdateDTO.scala (hunk 3 above)
Comment (mario-galic) [reply identity/app/idapiclient/UserUpdateDTO.scala]: Removed.

**Comment by mario-galic** [comment]
Closing in favour of https://github.com/guardian/frontend/pull/18103

**Comment by mario-galic** [comment]
Re-opened and put behind a switch by https://github.com/guardian/frontend/pull/18164
